iree/compiler/Dialect/Vulkan/Utils/TargetEnvUtils.cpp

```diff
@@ -100,22 +151,69 @@ spirv::ResourceLimitsAttr convertResourceLimits(
   MLIRContext *context = vkTargetEnv.getContext();
   auto vkCapabilities = vkTargetEnv.getCapabilitiesAttr();
   return spirv::ResourceLimitsAttr::get(
-      /*vendor_id=*/nullptr,
-      /*device_id=*/nullptr,
-      /*device_type=*/nullptr,
-      /*max_compute_shared_memory_size=*/nullptr,
+      vkCapabilities.maxComputeSharedMemorySize(),
       vkCapabilities.maxComputeWorkGroupInvocations(),
-      vkCapabilities.maxComputeWorkGroupSize(),
-      /*subgroup_size=*/nullptr, context);
+      vkCapabilities.maxComputeWorkGroupSize(), vkCapabilities.subgroupSize(),
+      context);
 }
 }  // anonymous namespace
 
-// TODO(antiagainst): register more SwiftShader extensions.
-const char *swiftShaderTargetEnvAssembly =
-    "#vk.target_env<v1.1, r(0), [VK_KHR_storage_buffer_storage_class], {"
-    "maxComputeWorkGroupInvocations = 128: i32, "
-    "maxComputeWorkGroupSize = dense<[128, 128, 64]>: vector<3xi32>"
-    "}>";
+// TODO(antiagainst): The following is good to get us started but it is
+// certainly not a scalable way to describe GPU targets. We need more proper
+// data structures and such.
+const char *getTargetEnvForTriple(llvm::StringRef triple) {
+  if (triple == "qualcomm-adreno640-unknown-android10") {
```

Comment:
This is hard-coding architectures. I am not sure that is very scalable. Could query vulkan API to get these?

Looking at this, I am actually not sure we want device ID and vendor name. If we could just use the Vulkan API to query things, then do we really need those? This is OK as a short/medium term solution. Ideally we shouldnt care?

Reply:
Regarding whether to just use Vulkan API to query, it actually depends. I view the CodeGen pipeline more as a "static" thing. It can compile towards all sorts of Vulkan targets; the current hardware on host (where we can query using Vulkan API) is just one of them. Think it as cross-compiling towards a different architecture in CPU land. We need such capabilities because in the future IREE needs to pack SPIR-V blobs for multiple Vulkan targets and not all of them are queryable with APIs on the current host. However, `#vk.target_env` is designed to be similar enough to what typically one query using the APIs (you can see the links embedded as comments pointing to the Vulkan structs). The goal is that it's very easy to construct such a target env by using information from `vulkaninfo` or vulkan database like https://vulkan.gpuinfo.org/.

Regarding the exact "triple", I've been going back and forth a while and settled down on the current scheme, tradeoffing verbosity and flexibility, and also trying to maintain some similarity with CPU side. The ultimate way is certainly to just provide the detailed `#vk.target_env`, which provide the most flexibility and detailed control. But It's very verbose. Most of the time we don't want to specify that every time; instead we just want some simpler "handles" to the hardware we are targeting. To specify a hardware, we need the vendor/architecture as a good label and to accommodate cases where we cannot get the information from the resource limits, like memory hierarchy. And the vendor/architecture isn't enough because SoC can tweak things and expose different functionalities depending on it. So that's the SoC/product field. OS also plays a role here because the same SoC with newer versions of OSes might have more capabilities. There are cases for this that I can point to. These fields aren't enough to exactly pin down all details for sure, but they are good approximation already and also caters to usability for us as developers. 

The representation here is certainly not scalable. We need proper data structures as written in the TODO in comments. But to get started, having a few strings is not that bad to me.

Reply:
Thanks @antiagainst for the clarification. Fine with landing this and iterating. Just for NVIDIA devices this list might become too big already. So something to keep in mind as this evolves.

Reply:
Yeah. At the moment each configuration is a separate string so it's combinational. We need to turn each component into its own enum and then parse the string instead, so we can probably treat ampere-rtx3080 and ampere-rtx3090 the same.